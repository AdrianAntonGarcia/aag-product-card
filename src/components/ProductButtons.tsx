import React from 'react';
import { CSSProperties, useCallback, useContext } from 'react';
import { ProductContext } from '../context';
import styles from '../styles/styles.module.css';

export interface Props {
  className?: string;
  style?: CSSProperties;
}
/**
 * Componente de los botones del producto
 */
export const ProductButtons = ({ className, style }: Props) => {
  // TODO: maxCount
  const { counter, increaseBy, maxCount } = useContext(ProductContext);
  const isMaxReached = useCallback(() => !!maxCount && maxCount === counter, [
    counter,
    maxCount,
  ]);
  return (
    <div style={style} className={`${styles.buttonsContainer} ${className}`}>
      <button
        className={styles.buttonMinus}
        onClick={() => {
          increaseBy(-1);
        }}
      >
        -
      </button>
      <div className={styles.countLabel}>{counter}</div>
      <button
        disabled={isMaxReached()}
        className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`}
        onClick={() => {
          increaseBy(1);
        }}
      >
        +
      </button>
    </div>
  );
};
