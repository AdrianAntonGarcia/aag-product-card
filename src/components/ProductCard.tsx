import React from 'react';
import { CSSProperties } from 'react';
import { useProduct } from '../hooks/useProduct';
import {
  InitialValues,
  OnChangeArgs,
  Product,
  ProductCardHandlers,
} from '../interfaces';
import { ProductContext } from '../context';
import styles from '../styles/styles.module.css';

export interface Props {
  product: Product;
  // children?: ReactElement | ReactElement[];
  children: (args: ProductCardHandlers) => JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: OnChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

/**
 * Componente que devuelve la tarjeta de un producto
 * @param param0
 * @returns
 */
export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const { Provider } = ProductContext;
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount,
      }}
    >
      <div style={style} className={`${styles.productCard} ${className}`}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          reset,
          increaseBy,
          product,
        })}
      </div>
    </Provider>
  );
};
