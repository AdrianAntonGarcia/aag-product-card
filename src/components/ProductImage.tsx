import React from 'react';
import { CSSProperties, useContext } from 'react';
import noImage from '../assets/no-image.jpg';
import { ProductContext } from '../context';
import styles from '../styles/styles.module.css';

export interface Props {
  img?: string;
  className?: string;
  style?: CSSProperties;
}
/**
 * Componente de la imagen del producto
 * @param param0
 * @returns
 */
export const ProductImage = ({ img = '', className, style }: Props) => {
  const { product } = useContext(ProductContext);
  let imageToShow;
  if (img) {
    imageToShow = img;
  } else if (product.img) {
    imageToShow = product.img;
  } else {
    imageToShow = noImage;
  }
  return (
    <img
      style={style}
      className={`${styles.productImg} ${className}`}
      src={imageToShow}
      alt="Product"
    />
  );
};
