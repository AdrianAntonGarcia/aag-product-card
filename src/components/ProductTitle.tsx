import React from 'react';
import { CSSProperties, useContext } from 'react';
import { ProductContext } from '../context';
import styles from '../styles/styles.module.css';

export interface Props {
  title?: string;
  className?: string;
  style?: CSSProperties;
}
/**
 * Componente del tÃ­tulo de la imagen
 * @param param0
 * @returns
 */
export const ProductTitle = ({ title, className, style }: Props) => {
  const { product } = useContext(ProductContext);
  return (
    <span style={style} className={`${styles.productDescription} ${className}`}>
      {title ? title : product.title}
    </span>
  );
};
