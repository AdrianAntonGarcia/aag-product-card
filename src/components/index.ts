import { ProductCard as ProductCardHOC } from './ProductCard';
import { ProductTitle, ProductImage, ProductButtons } from '.';
import { ProductCardHOCProps } from '../interfaces';

// export * from './ProductCard';
export { ProductImage } from './ProductImage';
export { ProductTitle } from './ProductTitle';
export { ProductButtons } from './ProductButtons';

export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
  Title: ProductTitle,
  Image: ProductImage,
  Buttons: ProductButtons,
});

export default ProductCard;
