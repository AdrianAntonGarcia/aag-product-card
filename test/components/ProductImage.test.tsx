import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard, ProductImage } from '../../src/components';
import { product2 } from '../data/products';

describe('ProductImage', () => {
  test('Debe mostrar el componente de la imagen pasada como parÃ¡metro', () => {
    const wrapper = renderer.create(
      <ProductImage img="./coffee-mug-prueba.png" />
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('debe de mostrar el componente con la imagen del producto', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>{() => <ProductImage />}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
