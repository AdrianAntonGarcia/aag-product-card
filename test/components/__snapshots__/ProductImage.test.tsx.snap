// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage Debe mostrar el componente de la imagen pasada como par√°metro 1`] = `
<img
  alt="Product"
  className="productImg undefined"
  src="./coffee-mug-prueba.png"
/>
`;

exports[`ProductImage debe de mostrar el componente con la imagen del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="./coffee-mug2.png"
  />
</div>
`;
